lib/SPIRV/libSPIRV/SPIRVEntry.h

```diff
@@ -665,19 +665,38 @@ class SPIRVExecutionMode : public SPIRVAnnotation<OpExecutionMode> {
 };
 
 class SPIRVComponentExecutionModes {
-  typedef std::map<SPIRVExecutionModeKind, SPIRVExecutionMode *>
+  typedef std::multimap<SPIRVExecutionModeKind, SPIRVExecutionMode *>
       SPIRVExecutionModeMap;
+  typedef std::pair<SPIRVExecutionModeMap::const_iterator,
+                    SPIRVExecutionModeMap::const_iterator>
+      SPIRVExecutionModeRange;
 
 public:
   void addExecutionMode(SPIRVExecutionMode *ExecMode) {
-    ExecModes[ExecMode->getExecutionMode()] = ExecMode;
+    SPIRVExecutionModeKind EMK = ExecMode->getExecutionMode();
+    // There should not be more than 1 execution mode kind except the ones
+    // mentioned in SPV_KHR_float_controls.
+#ifndef NDEBUG
+    bool IsValid =
+        (ExecModes.count(EMK) == 0 || EMK == ExecutionModeDenormPreserve ||
```

Comment:
From the [extension spec](http://htmlpreview.github.io/?https://github.com/KhronosGroup/SPIRV-Registry/blob/master/extensions/KHR/SPV_KHR_float_controls.html):

> An OpEntryPoint can set at most one of the DenormFlushToZero, DenormPreserve Execution Modes for the given Floating-Point Type Bit Width.
>
> An OpEntryPoint can set at most one of the RoundingModeRTE, RoundingModeRTZ Execution Modes for the given Floating-Point Type Bit Width.

If I understand correctly, there are a few cases missing in this validation: for example, there is no check that `ExecutionModeRoundingModeRTE` is set together with `ExecutionModeRoundingModeRTZ` for the same `Target Width`.

This is probably a not critical comment since this code is anyway under `#ifndef`, but it worth adding a `TODO`, at least

Reply:
Fixed